src/corelibs/U2View/src/ov_msa/exclude_list/MsaExcludeList.cpp

```diff
@@ -0,0 +1,593 @@
+
+    namesAndSequenceSplitter->addWidget(nameListView);
+    namesAndSequenceSplitter->setStretchFactor(0, 1);
+    namesAndSequenceSplitter->setCollapsible(0, false);
+    namesAndSequenceSplitter->addWidget(sequenceView);
+    namesAndSequenceSplitter->setStretchFactor(1, 3);
+    namesAndSequenceSplitter->setCollapsible(1, false);
+
+    this->setContextMenuPolicy(Qt::PreventContextMenu);
+    nameListView->setContextMenuPolicy(Qt::CustomContextMenu);
+
+    moveToMsaAction = new QAction(tr("Move to alignment"), this);
+    moveToMsaAction->setToolTip(tr("Move selected Exclude List sequences to MSA"));
+    moveToMsaAction->setIcon(QIcon(":core/images/arrow-move-up.png"));
+    connect(moveToMsaAction, &QAction::triggered, this, &MsaExcludeListWidget::moveExcludeListSelectionToMaObject);
+
+    auto moveToMsaButton = new QToolButton();
+    moveToMsaButton->setObjectName("exclude_list_move_to_msa_button");
+    moveToMsaButton->setDefaultAction(moveToMsaAction);
+    moveToMsaButton->setToolButtonStyle(Qt::ToolButtonIconOnly);
+    toolbarLayout->addWidget(moveToMsaButton);
+    toolbarLayout->addStrut(10);
+
+    auto moveFromMsaAction = viewContext->getMoveMsaSelectionToExcludeListAction(msaEditor);
+    if (moveFromMsaAction != nullptr) {
+        auto moveFromMsaButton = new QToolButton();
+        moveFromMsaButton->setObjectName("exclude_list_move_to_exclude_list_button");
+        moveFromMsaButton->setDefaultAction(moveFromMsaAction);
+        moveFromMsaButton->setToolButtonStyle(Qt::ToolButtonIconOnly);
+        toolbarLayout->addWidget(moveFromMsaButton);
+        toolbarLayout->addStrut(10);
+    }
+
+    auto msaObject = msaEditor->getMaObject();
+    connect(msaObject, &MultipleSequenceAlignmentObject::si_alignmentChanged, this, &MsaExcludeListWidget::handleUndoRedoInMsaEditor);
+
+    excludeListFilePath = msaEditor->property(PROPERTY_LAST_USED_EXCLUDE_LIST_FILE).toString();
+    if (excludeListFilePath.isEmpty() || !QFileInfo::exists(excludeListFilePath)) {
+        GUrl msaUrl = msaObject->getDocument()->getURL();
+        excludeListFilePath = msaUrl.dirPath() + "/" + msaUrl.baseFileName() + "." + EXCLUDE_LIST_FILE_SUFFIX;
+        if (GUrl(excludeListFilePath) == msaUrl) {
+            excludeListFilePath = msaUrl.getURLString() + "." + EXCLUDE_LIST_FILE_SUFFIX;
+        }
+    }
+
+    connect(AppContext::getTaskScheduler(), &TaskScheduler::si_stateChanged, this, &MsaExcludeListWidget::trackMsaObjectSaveTask);
+
+    toolbarLayout->addStretch();
+
+    loadExcludeList(true);
+}
+
+MsaExcludeListWidget::~MsaExcludeListWidget() {
+    if (isDirty) {
+        runSaveTask(excludeListFilePath);
+    }
+}
+
+int MsaExcludeListWidget::addEntry(const DNASequence &sequence, int excludeListRowId) {
+    SAFE_POINT(sequence.alphabet != nullptr, "Sequence must be fully defined!", 0);  // By default MSA row sequences have no alphabet. Catch this kind of error ASAP.
```

Comment:
`L10n`

Reply:
Please forget about L10n in SAFE_POINTS. Such code will not pass a code reviews.